src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1150,4 +1229,176 @@ _mongoc_sasl_prep (const char *in_utf8, int in_utf8_len, bson_error_t *err)
    return bson_strdup (in_utf8);
 #endif
 }
+
+int
+_mongoc_utf8_char_length (const char *c)
+{
+   int length;
+
+   // UTF-8 characters are either 1, 2, 3, or 4 bytes and the character length
+   // can be determined by the first byte
+   if ((*c & 0x80) == 0)
+      length = 1;
+   else if ((*c & 0xe0) == 0xc0)
+      length = 2;
+   else if ((*c & 0xf0) == 0xe0)
+      length = 3;
+   else if ((*c & 0xf8) == 0xf0)
+      length = 4;
+   else
+      length = 1;
+
+   return length;
+}
+
+int
+_mongoc_utf8_string_length (const char *s)
+{
+   const char *c = s;
+
+   int str_length = 0;
+   int utf8_char_length;
+
+   while (*c) {
+      utf8_char_length = _mongoc_utf8_char_length (c);
+
+      if (!_mongoc_utf8_is_valid (c, utf8_char_length))
+         return -1;
+
+      str_length++;
+      c += utf8_char_length;
+   }
+
+   return str_length;
+}
+
+
+bool
+_mongoc_utf8_is_valid (const char *c, int length)
+{
```

Comment:
```suggestion
{
   BSON_ASSERT_PARAM (c);

```
Suggest asserting `BSON_ASSERT_PARAM (c);`. It does not make sense to request the length of a not-string (+ `strlen()` consistency).